### DISCUSSION TITLE
Kratos as identityt server and hydra as opn.Id/0auth provider guide

### DISCUSSION THREAD

**Discussion by tima-t** [ory/kratos#1656] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
My problem is that I cannot connect ory kratos (latest) to hydra (latest) using https://github.com/ory/kratos-selfservice-ui-node/ 

I've tried to look at the branch hydra-integration-2021 and to integrate kratos(latest version) and hydra via kratos-selfservice-ui-node, but I cannot finalzie the integration for two days now. Both document look well documented, but as they are big apps by their own it is hard to me to connect them and make them work together. I am searhcing for godo examples but there is not a clear step by step guide how to connect these two services. 
I saw that it will eb a natiev itnegratiion between hydra and kratos but it is still WIP.  https://www.ory.sh/kratos/docs/concepts/federation/
Is it going to be live soon, or can I look and even test the future branch if it is ready for beta? 

**Describe the solution you'd like**

A docker setup that runs ory kratos as identity provider and hydra as open.Id/ 0Auth2 provider. 
If that is not possible or still wip, I would like a clear step by step guide how to implement and add this feature in kratos-selfservice-ui-node, because this version https://github.com/ory/kratos-selfservice-ui-node/tree/hydra-integration-2021 is already outdated and the migration to latest versions  is not headless

**Describe alternatives you've considered**
I would like a clear step by step guide how to implement and add ory/hydra  in kratos-selfservice-ui-node

**Comment by aeneasr** [comment]
Welcome, and sorry to hear that this is complicated! We have an issue tracking this #273 - it is currently tracked for v0.9.0 but depending on how it goes (the milestone is pretty large) it could also be pushed back to v0.10.0. My recommendation would be to start with Ory Kratos first, and then add Ory Hydra in at a later stage. This will make you comfortable with Ory Kratos! And depending on what you want to do will also take a bit of time.

Once you have figured out how to best use Ory Kratos, the linked repo might make more sense and will be easier to adjust!

**Comment by atreya2011** [comment, marked as answer]
@tima-t
I have created an extremely simple implementation here in Go. Should be easy to follow and translate to any other language of your choice.
Its WIP and does not have a README at the moment. Planning to add instructions of how to spin everything up within the next few days,
and add verification, recovery flows as well!

https://github.com/atreya2011/go-kratos-test `without hydra`
https://github.com/atreya2011/go-kratos-test/tree/hydra `with hydra`

Diff
https://github.com/atreya2011/go-kratos-test/compare/hydra

All I did was follow all the UML diagrams in the docs and the reference implementations in Node.
And as @aeneasr suggested, I first started with Kratos, got everything working and then moved on to integrating Hydra.
Kratos is an extremely well thought out and thoroughly documented Identity Provider! Thanks for this @aeneasr 🙇🏼‍♂️

**Reply by azazel75** [reply]
@atreya2011 Are you sure the hydra version works? The compose file mentions an `auth-provider` service but it isn't in those being spawn!?

**Reply by atreya2011** [reply]
@azazel75 Ah sorry for not updating the README yet, but, yes, I confirmed that hydra works. I have updated things a bit locally and pushed them just now.

I made a frontend UI using [Next.js](https://nextjs.org/) + [Next-Auth](https://next-auth.js.org/) and added Kratos-Hydra combo as a custom provider and confirmed that I got the access_token and the id_token.

Here is the Frontend implementation. https://github.com/atreya2011/next-auth-kratos-hydra-test
It is also WIP. I am planning to update the README for both repos in the next few days 🙏🏼

**Reply by atreya2011** [reply]
Here is the GIF btw
![Peek 2021-11-18 08-33](https://user-images.githubusercontent.com/44151328/142299596-74a03f44-4187-457b-a8d7-8dfee075349b.gif)

**Reply by AlimByWater** [reply]
I see you did it with browser flow, so my question is how to make it on server-side?
For example, how to handle cookie during the registration? All i see in docs is warnings like that 

> Never use API flows to implement Browser applications! Using API flows in Single-Page-Apps as well as server-side apps opens up several potential attack vectors, including Login and other CSRF attacks.

**Reply by atreya2011** [reply]
@AlimByWater 
To be honest, I haven't tried the API flows yet, and my understanding is that you have to manually handle cookies for API flows which kratos does for you if you use the browser flows.
I have to yet to figure out a secure way to implement API flows but an official example would be very helpful indeed.

**Reply by vinckr** [reply]
This is great @atreya2011 !
Would you be up to creating a dedicated **show&tell** discussion for it e.g. "Kratos & Hydra Integration"? 
Then we can pin it and give it the attention it deserves 🙏 
Lots of folks have been looking for something like this 👍

**Reply by AlimByWater** [reply]
@atreya2011
I guess api registrations(for example) endpoint can incapsulate all proccess, like so: initialize registration flow -> submite registration -> return session token. Do u see any flaws in this method?

**Reply by atreya2011** [reply]
@vinckr

> Would you be up to creating a dedicated show&tell discussion for it e.g. "Kratos & Hydra Integration"?
Then we can pin it and give it the attention it deserves pray
Lots of folks have been looking for something like this +1

Thank you for the opportunity! Please let me how to create a dedicated **show&tell** discussion for it 🙇🏼‍♂️ 
Is it okay if I just create a new discussion in GitHub or is there a community forum section for posting a new **show&tell** discussion? 

However, I still have to address the following issues:
- Consent UI Implementation
- User breaks login flow if they try to register a new account and gets back to the login screen (keep a hydra login flow state like the reference Node.js implementation)

**Reply by atreya2011** [reply]
@AlimByWater 

> I guess api registrations(for example) endpoint can incapsulate all proccess, like so: initialize registration flow -> submite registration -> return session token. Do u see any flaws in this method?

Yes I saw a very similar implementation over here and it is possible. 
However, the only thing missing is the security protection we get by means of csrf tokens. 

EDIT: If we employ our own CSRF protection in our web servers and store session details server-side, I don't see any reason for NOT using API flows instead of the browser flows.

https://github.com/ory/kratos/blob/master/examples/go/selfservice/registration/main.go

**Reply by vinckr** [reply]
Hey @atreya2011 
Yes, just go to https://github.com/ory/hydra/discussions/new and pick "Show & Tell" as category :-).

And don't worry if your project is not 100% done or some things are missing, that is totally ok. The show & tell post will hopefully also attract some more contributors.

**Reply by atreya2011** [reply]
@vinckr Thanks for the clarification :) I will create a dedicated show&tell discussion within the next few days 🙏🏼 
And thank you for featuring my sample implementation over here: https://github.com/ory/hydra/discussions/2855#discussioncomment-1670601 🙇🏼‍♂️

**Reply by atreya2011** [reply]
@vinckr Sorry for the delay. I have created a dedicated show&tell discussion :)
https://github.com/ory/hydra/discussions/2873

**Reply by vinckr** [reply]
Thanks Atreya! ❤️

**Comment by vinckr** [comment]
Hey @azazel75 @AlimByWater @tima-t 
The Ory Kratos and Ory Hydra integration just landed with Hydra 2.0, 
see the [release notes](https://github.com/ory/hydra/releases/tag/v2.0.0) for more details!
